presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,521 @@
+        }
+
+        @Override
+        protected FilterNode rewrite(Context context, FilterNode node, BiMap<DereferenceExpression, Symbol> expressions)
+        {
+            PlanNode source = node.getSource();
+            Assignments assignments = Assignments.builder().putIdentities(source.getOutputSymbols()).putAll(expressions.inverse()).build();
+            ProjectNode projectNode = new ProjectNode(context.getIdAllocator().getNextId(), source, assignments);
+
+            return new FilterNode(
+                    context.getIdAllocator().getNextId(),
+                    projectNode,
+                    ExpressionTreeRewriter.rewriteWith(new PushDownDereferences.DereferenceReplacer(expressions), node.getPredicate()));
+        }
+    }
+
+    class ExtractFromJoin
+            extends ExtractProjectDereferences<JoinNode>
+    {
+        ExtractFromJoin()
+        {
+            super(JoinNode.class);
+        }
+
+        @Override
+        protected JoinNode rewrite(Context context, JoinNode joinNode, BiMap<DereferenceExpression, Symbol> expressions)
+        {
+            Assignments.Builder leftSideDereferences = Assignments.builder();
+            Assignments.Builder rightSideDereferences = Assignments.builder();
+
+            for (Map.Entry<Symbol, DereferenceExpression> entry : expressions.inverse().entrySet()) {
+                Symbol baseSymbol = getBase(entry.getValue());
+                if (joinNode.getLeft().getOutputSymbols().contains(baseSymbol)) {
+                    leftSideDereferences.put(entry.getKey(), entry.getValue());
+                }
+                else {
+                    rightSideDereferences.put(entry.getKey(), entry.getValue());
+                }
+            }
+            PlanNode leftNode = createProjectBelow(joinNode.getLeft(), leftSideDereferences.build(), context.getIdAllocator());
+            PlanNode rightNode = createProjectBelow(joinNode.getRight(), rightSideDereferences.build(), context.getIdAllocator());
+
+            return new JoinNode(
+                    context.getIdAllocator().getNextId(),
+                    joinNode.getType(),
+                    leftNode,
+                    rightNode,
+                    joinNode.getCriteria(),
+                    ImmutableList.<Symbol>builder().addAll(leftNode.getOutputSymbols()).addAll(rightNode.getOutputSymbols()).build(),
+                    joinNode.getFilter().map(expression -> ExpressionTreeRewriter.rewriteWith(new PushDownDereferences.DereferenceReplacer(expressions), expression)),
+                    joinNode.getLeftHashSymbol(),
+                    joinNode.getRightHashSymbol(),
+                    joinNode.getDistributionType(),
+                    joinNode.isSpillable(),
+                    joinNode.getDynamicFilters());
+        }
+    }
+
+    /**
+     * Push down dereferences from ProjectNode to child nodes if possible
+     */
+    private abstract class PushdownDereferencesInProject<N extends PlanNode>
```

Comment:
Relying inheritance to reuse behavior makes the code harder to follow and reason about (is it just about code reuse? Is there shared state that's important? etc). It'd be better to model it via some form of delegation. E.g., The rule can take another object that performs the job of `pushDownDereferences`.